### PR TITLE
fix: do not assume latest DatasetVersion is active

### PR DESCRIPTION
**Pull request by danieljhegeman**
Change relevant db provider methods to strictly rely on datasets associated with latest CollectionVersion as the 'active' DatasetVersion for a given DatasetId. This fixes a bug introduced by #5355 which allows rollback to previous DatasetVersions without deleting the most-recently-created DatasetVersion row.

## Reason for Change

- #4811 

## Changes

- replace usage of `get_all_versions_for_dataset` to find latest (active) DatasetVersion with a new db provider method `get_most_recent_active_dataset_version`.

## Testing steps

- unit tested

### CODE REVIEW

File: backend/layers/persistence/persistence.py [hunk 1]
```diff
@@ -596,23 +596,35 @@ def get_dataset_version(self, dataset_version_id: DatasetVersionId, get_tombston
                     return None
             return self._hydrate_dataset_version(dataset_version)
 
+    def get_most_recent_active_dataset_version(self, dataset_id: DatasetId) -> Optional[DatasetVersion]:
+        """
+        Returns the must recently active Dataset version for a canonical dataset_id
```
Comment (Bento007) [inline backend/layers/persistence/persistence.py:601]:
```suggestion
        Returns the most recently active Dataset version for a canonical dataset_id
```

What does it mean if it returns None?

File: backend/layers/persistence/persistence_interface.py [hunk 2]
```diff
@@ -138,6 +138,11 @@ def get_dataset_version(self, dataset_version_id: DatasetVersionId, get_tombston
         Returns a dataset version by id.
         """
 
+    def get_most_recent_active_dataset_version(self, dataset_id: DatasetId) -> Optional[DatasetVersion]:
+        """
+        Returns the must recent, active Dataset version for a canonical dataset_id
```
Comment (Bento007) [inline backend/layers/persistence/persistence_interface.py:143]:
```suggestion
        Returns the most recent, active Dataset version for a canonical dataset_id
```

File: backend/layers/persistence/persistence_mock.py [hunk 3]
```diff
@@ -322,7 +322,32 @@ def get_dataset_versions_by_id(
             dataset_versions.append(dataset_version)
         return dataset_versions
 
+    def get_most_recent_active_dataset_version(self, dataset_id: DatasetId) -> Optional[DatasetVersion]:
+        """
+        Returns the must recent, active Dataset version for a canonical dataset_id
```
Comment (Bento007) [inline backend/layers/persistence/persistence_mock.py:327]:
```suggestion
        Returns the most recent, active Dataset version for a canonical dataset_id
```

**Review by Bento007 — APPROVED** [review]
A few typos to fix

File: backend/layers/persistence/persistence.py (hunk 1 above)
Comment (danieljhegeman) [reply backend/layers/persistence/persistence.py:601]: No such canonical Dataset. Similar to [the way this method works](https://github.com/chanzuckerberg/single-cell-data-portal/blob/d09ddc97937aad4a9a345c0c12dcc2efb7fdc0d0/backend/layers/persistence/persistence.py#L185-L187).

**Pull request merged**